ouroboros-praos/src/consensus/mod.rs

```diff
@@ -0,0 +1,349 @@
+static CERTIFIED_NATURAL_MAX: LazyLock<FixedDecimal> = LazyLock::new(|| {
+    FixedDecimal::from_str(
+        "1157920892373161954235709850086879078532699846656405640394575840079131296399360000000000000000000000000000000000",
+        34,
+    )
+        .expect("Infallible")
+});
+
+/// Validator for a block using praos consensus.
+pub struct BlockValidator<'b> {
+    header: &'b babbage::Header,
+    pool_info: &'b dyn PoolInfo,
+    epoch_nonce: &'b Hash<32>,
+    // c is the ln(1-active_slots_coeff). Usually ln(1-0.05)
+    c: &'b FixedDecimal,
+}
+
+impl<'b> BlockValidator<'b> {
+    // TODO: This function is currently only used in tests. Remove warning suppression in the future
+    #[allow(unused)]
+    fn new(
+        header: &'b babbage::Header,
+        pool_info: &'b dyn PoolInfo,
+        epoch_nonce: &'b Hash<32>,
+        c: &'b FixedDecimal,
+    ) -> Self {
+        Self {
+            header,
+            pool_info,
+            epoch_nonce,
+            c,
+        }
+    }
+
+    fn validate_babbage_compatible(&self) -> bool {
+        let span = span!(tracing::Level::TRACE, "validate_babbage_compatible");
+        let _enter = span.enter();
+
+        // Grab all the values we need to validate the block
+        let issuer_vkey = &self.header.header_body.issuer_vkey;
+        let pool_id: PoolId = issuer_vkey_to_pool_id(issuer_vkey);
+        let vrf_vkey: &[u8; VRF_PUBLIC_KEY_SIZE] = self
+            .header
+            .header_body
+            .vrf_vkey
+            .as_slice()
+            .try_into()
+            .expect("Infallible");
+        if !self.ledger_matches_block_vrf_key_hash(&pool_id, vrf_vkey) {
+            // Fail fast if the vrf key hash in the block does not match the ledger
+            return false;
+        }
+        let sigma: FixedDecimal = match self.pool_info.sigma(&pool_id) {
+            Ok(sigma) => {
+                FixedDecimal::from(sigma.numerator) / FixedDecimal::from(sigma.denominator)
+            }
+            Err(error) => {
+                warn!("{:?} - {:?}", error, pool_id);
+                return false;
+            }
+        };
+        let absolute_slot = self.header.header_body.slot;
+
+        // Calculate the leader VRF output hash from the block vrf result
+        let mut hasher = Hasher::<256>::new();
+        hasher.input(&[0x4C_u8]); /* "L" */
+        hasher.input(&self.header.header_body.vrf_result.0);
+        let leader_vrf_output = hasher.finalize();
+
+        let block_vrf_proof_hash: [u8; VRF_PROOF_HASH_SIZE] = self
+            .header
+            .header_body
+            .vrf_result
+            .0
+            .as_slice()
+            .try_into()
+            .expect("Infallible");
+        let block_vrf_proof: [u8; VRF_PROOF_SIZE] = self
+            .header
+            .header_body
+            .vrf_result
+            .1
+            .as_slice()
+            .try_into()
+            .expect("Infallible");
```

Comment:
This screams for a better interface in pallas-primitives.

Reply:
A better api exists in pallas-traverse, but we decided we wanted to not rely on that particular crate. I only pulled it in so I could make a test work.

Reply:
See: https://github.com/txpipe/pallas/pull/517